### PR TITLE
gnrc_rpl: port to nib

### PR DESCRIPTION
**Pull request by cgundogan**
replaces fib with nib/ft

WIP for now. There are still some modifications missing in the tests for `gnrc_ipv6_nib_ft_add` e.g.
Also, I have to re-evaluate the netif2 changes ..
Will continue tomorrow with that.

This PR is part of the network layer remodelling effort:
PR dependencies

depends on #8083

### CODE REVIEW

File: sys/include/net/gnrc/ipv6/nib/ft.h [hunk 1]
```diff
@@ -68,14 +68,16 @@ int gnrc_ipv6_nib_ft_get(const ipv6_addr_t *dst, gnrc_pktsnip_t *pkt,
...
+ * @param[in] ltime     Lifetime in seconds of this route entry.
  * @param[in] iface     The interface to @p next_hop. May not be 0.
  *
  * @return  0, on success.
  * @return  -EINVAL, if a parameter was of invalid value.
  * @return  -ENOMEM, if there was no space left in forwarding table.
  */
 int gnrc_ipv6_nib_ft_add(const ipv6_addr_t *dst, unsigned dst_len,
-                         const ipv6_addr_t *next_hop, unsigned iface);
+                         const ipv6_addr_t *next_hop, uint16_t ltime,
```
Comment (miri64) [inline sys/include/net/gnrc/ipv6/nib/ft.h]: This was kept out of the intentionally. The routing protocol should timeout the routes externally (e.g. by storing the context for the timer in the protocol's routing table)

Comment (cgundogan) [reply sys/include/net/gnrc/ipv6/nib/ft.h]: This is weird. Every FIB entry should have a lifetime. RPL is not aware of link failures way down the tree as it does not maintain a children list, only a parent list. So RPL does not know when to prune an entry in that case. FIB maintenance should be in the FIB, not in a routing protocol.

Comment (miri64) [reply sys/include/net/gnrc/ipv6/nib/ft.h]: This isn't about FIB maintenance, but about route maintenance IMHO. The lifetime of the route (and its scale) is intrinsic to the routing protocol (if you want a proof for that, see various discussions in our GitHub issue history about which unit should be used with route lifetimes).

Comment (miri64) [reply sys/include/net/gnrc/ipv6/nib/ft.h]: Iff we do it like this I have some questions: Why is the route lifetime in seconds (isn't milliseconds a more usual unit for that)? Why did you choose `uint16_t` as the type for that?

File: sys/net/gnrc/routing/rpl/gnrc_rpl_control_messages.c [hunk 2]
```diff
@@ -754,60 +742,28 @@ void gnrc_rpl_send_DAO(gnrc_rpl_instance_t *inst, ipv6_addr_t *destination, uint
...
-                      (int) prefix_length);
-
-                if ((*ptr = _dao_target_build(*ptr, addr, (uint8_t) prefix_length)) == NULL) {
-                    DEBUG("RPL: Send DAO - no space left in packet buffer\n");
-                    mutex_unlock(&(gnrc_ipv6_fib_table.mtx_access));
-                    return;
-                }
-            }
+    /* add external and RPL FT entries */
+    /* TODO: nib: dropped support for external transit options for now */
```
Comment (miri64) [inline sys/net/gnrc/routing/rpl/gnrc_rpl_control_messages.c]: What do you mean by that?

File: sys/net/gnrc/network_layer/ipv6/nib/nib_ft.c [hunk 3]
```diff
@@ -45,12 +46,20 @@ int gnrc_ipv6_nib_ft_add(const ipv6_addr_t *dst, unsigned dst_len,
...
+            ptr = (void *) dr;
+        }
     }
 #if GNRC_IPV6_NIB_CONF_ROUTER
     else {
         dst_len = (dst_len > 128) ? 128 : dst_len;
-        ptr = _nib_ft_add(next_hop, iface, dst, dst_len);
+        _nib_offl_entry_t *offl = _nib_ft_add(next_hop, iface, dst, dst_len);
+        if (offl) {
+            offl->valid_until = offl->pref_until = ltime * MS_PER_SEC;
```
Comment (miri64) [inline sys/net/gnrc/network_layer/ipv6/nib/nib_ft.c]: Iff we add this feature we need to add a new field to `_nib_offl_entry_t`. You are risking here to override prefix lifetimes and thus causing potential side effects.

File: sys/net/gnrc/routing/rpl/gnrc_rpl_dodag.c [hunk 4]
```diff
@@ -216,22 +216,15 @@ bool gnrc_rpl_parent_remove(gnrc_rpl_parent_t *parent)
     gnrc_rpl_dodag_t *dodag = parent->dodag;
 
     if (parent == dodag->parents) {
-        fib_remove_entry(&gnrc_ipv6_fib_table,
-                         (uint8_t *) ipv6_addr_unspecified.u8,
-                         sizeof(ipv6_addr_t));
+        gnrc_ipv6_nib_ft_del(&ipv6_addr_unspecified, 0);
```
Comment (miri64) [inline sys/net/gnrc/routing/rpl/gnrc_rpl_dodag.c]: Note to self: this might cause problems with address registration => investigate.

File: sys/include/net/gnrc/ipv6/nib/ft.h (hunk 1 above)
Comment (cgundogan) [reply sys/include/net/gnrc/ipv6/nib/ft.h]: >Why is the route lifetime in seconds (isn't milliseconds a more usual unit for that)? Why did you choose uint16_t as the type for that?

We had the discussion for `gnrc_fib` that milliseconds granularity is overkill and the port to deciseconds was never done.. I guess having a second granularity is okay for our constrained IoT scenario and routing protocols usually are not that fast in noticing routing state changes.

Comment (cgundogan) [reply sys/include/net/gnrc/ipv6/nib/ft.h]: Oh, and I chose `uint16_t` because that's what the default router struct is also using.

**Comment by miri64** [comment]
Ideally there shouldn't be any comment here. But just in case: this is a warning for that ;-).

**Comment by miri64** [comment]
Worked :-)

File: sys/net/gnrc/routing/rpl/gnrc_rpl_control_messages.c (hunk 2 above)
Comment (miri64) [reply sys/net/gnrc/routing/rpl/gnrc_rpl_control_messages.c]: In case we don't get this in before your vacation and someone else needs to take over: can you please answer my question?

File: sys/net/gnrc/routing/rpl/gnrc_rpl_control_messages.c [hunk 5]
```diff
@@ -754,60 +742,28 @@ void gnrc_rpl_send_DAO(gnrc_rpl_instance_t *inst, ipv6_addr_t *destination, uint
...
+    void *ft_state = NULL;
+    gnrc_ipv6_nib_ft_t fte;
+    while(gnrc_ipv6_nib_ft_iter(NULL, dodag->iface, &state, &fte)) {
```
Comment (miri64) [inline sys/net/gnrc/routing/rpl/gnrc_rpl_control_messages.c]: Should be `ft_state` right?

File: sys/net/gnrc/routing/rpl/gnrc_rpl_control_messages.c (hunk 2 above)
Comment (cgundogan) [reply sys/net/gnrc/routing/rpl/gnrc_rpl_control_messages.c]: RPL is able to distribute routing states that were not learned/produced by RPL, but from an external source – like another routing protocol or a manually admnistered route entry. These routing states should be marked with the `E` flag in DAO transit optoins [1].
>   External (E): 1-bit flag.  The 'E' flag is set to indicate that the
         parent router redistributes external targets into the RPL
         network.  An external Target is a Target that has been learned
         through an alternate protocol.  The external targets are listed
         in the Target options that immediately precede the Transit
         Information option.  An external Target is not expected to
         support RPL messages and options.

In the "old" fib implementation user defined flags could be stored along the forwarding entry. We used that to mark a forwarding entry as "learned by RPL" and used this information to decide whether to set or not set the `E` flag in DAOs.

This feature is missing in the current `nib_ft` implementation that's why I put a note about dropped support there.

[1] https://tools.ietf.org/html/rfc6550#section-6.7.8

**Comment by cgundogan** [comment]
rebased onto #8083

File: sys/net/gnrc/netif/gnrc_netif.c [hunk 6]
```diff
@@ -1216,7 +1216,7 @@ static void *_gnrc_netif_thread(void *args)
                 DEBUG("gnrc_netif: GNRC_NETAPI_MSG_TYPE_SET received. opt=%s\n",
                       netopt2str(opt->opt));
 #else
-                DEBUG("gnrc_netif: GNRC_NETAPI_MSG_TYPE_SET received. opt=%s\n",
+                DEBUG("gnrc_netif: GNRC_NETAPI_MSG_TYPE_SET received. opt=%d\n",
```
Comment (miri64) [inline sys/net/gnrc/netif/gnrc_netif.c]: Unrelated, but okay, as long as it stays its own commit.

File: sys/net/gnrc/network_layer/ipv6/nib/_nib-arsm.c [hunk 7]
```diff
@@ -281,7 +281,7 @@ void _handle_snd_ns(_nib_onl_entry_t *nbr)
                 _set_nud_state(netif, nbr,
                                GNRC_IPV6_NIB_NC_INFO_NUD_STATE_UNREACHABLE);
             }
-            /* falls through intentionally */
+            /* intentionally falls through */
```
Comment (miri64) [inline sys/net/gnrc/network_layer/ipv6/nib/_nib-arsm.c]: Dito (also has GCC become so picky??!!? oO)

File: sys/net/gnrc/routing/rpl/gnrc_rpl_control_messages.c [hunk 8]
```diff
@@ -420,22 +418,13 @@ bool _parse_options(int msg_type, gnrc_rpl_instance_t *inst, gnrc_rpl_opt_t *opt
...
-                fib_add_entry(&gnrc_ipv6_fib_table, dodag->iface, target->target.u8,
-                              sizeof(ipv6_addr_t), fib_dst_flags, src->u8,
-                              sizeof(ipv6_addr_t), FIB_FLAG_RPL_ROUTE,
-                              (dodag->default_lifetime * dodag->lifetime_unit) *
-                              MS_PER_SEC);
+                      target->prefix_length);
+
+                gnrc_ipv6_nib_ft_add(&(target->target), target->prefix_length, src,
+                                     dodag->iface,
+                                     dodag->default_lifetime * dodag->lifetime_unit);
```
Comment (miri64) [inline sys/net/gnrc/routing/rpl/gnrc_rpl_control_messages.c]: Not part of this PR, but is it really necessary to store the `lifetime_unit` of a DODAG Oo?

File: sys/net/gnrc/routing/rpl/gnrc_rpl_control_messages.c (hunk 2 above)
Comment (miri64) [reply sys/net/gnrc/routing/rpl/gnrc_rpl_control_messages.c]: Okay. I guess we'll need to extend the NIB for something like this later on ;-).

File: sys/net/gnrc/network_layer/ipv6/nib/_nib-arsm.c (hunk 7 above)
Comment (cgundogan) [reply sys/net/gnrc/network_layer/ipv6/nib/_nib-arsm.c]: Don't know if we have a stricter fall through level now, but assuming we have `level 3`[1],
then the previous string doesn't match the given regex. However, my compiler was able to compile it sometimes, and other times it didn't want to .. with the provided change it worked every time so far.

[1] https://gcc.gnu.org/onlinedocs/gcc/Warning-Options.html

File: sys/net/gnrc/routing/rpl/p2p/gnrc_rpl_p2p.c [hunk 9]
```diff
@@ -346,11 +346,9 @@ void gnrc_rpl_p2p_recv_DRO(gnrc_pktsnip_t *pkt, ipv6_addr_t *src)
     }
 
     if (gnrc_ipv6_netif_find_by_addr(&me, &addr) == dodag->iface) {
```
Comment (miri64) [inline sys/net/gnrc/routing/rpl/p2p/gnrc_rpl_p2p.c]: Are you sure, this is the right function as well?

File: sys/net/gnrc/routing/rpl/gnrc_rpl_control_messages.c (hunk 8 above)
Comment (cgundogan) [reply sys/net/gnrc/routing/rpl/gnrc_rpl_control_messages.c]: The lifetime and the lifetimeunit are both distributed in DIOs. Need to store that to further distribute it (:

Comment (miri64) [reply sys/net/gnrc/routing/rpl/gnrc_rpl_control_messages.c]: Ok! :-)

File: sys/net/gnrc/routing/rpl/p2p/gnrc_rpl_p2p.c (hunk 9 above)
Comment (miri64) [reply sys/net/gnrc/routing/rpl/p2p/gnrc_rpl_p2p.c]: Discussed.

**Comment by miri64** [comment]
Can you also please squash 70d0164 and 9873671 They belong together IMHO.

**Comment by cgundogan** [comment]
>Can you also please squash 70d0164 and 9873671 They belong together IMHO.

done

**Comment by miri64** [comment]
Go!

**Comment by cgundogan** [comment]
woohoo
